File: core/src/net/brenig/pixelescape/game/entity/EntityMoving.java
package net.brenig.pixelescape.game.entity;

import net.brenig.pixelescape.game.data.GameMode;

public abstract class EntityMoving extends Entity {

	protected float xVel, yVel;

	protected void checkMaxVelocity(GameMode gameMode) {
		xVel = Math.min(xVel, gameMode.getMaxEntitySpeed());
		yVel = Math.min(yVel, gameMode.getMaxEntitySpeed());
	}

	public void setVelocity(float xVel, float yVel) {
		this.xVel = xVel;
		this.yVel = yVel;
	}

	/**
	 * moves the entity
	 * <p>
	 * note: does note check velocity limits
	 *
	 * @param delta time passed
	 */
	protected void move(float delta) {
		xPos += xVel * delta;
		yPos += yVel * delta;
	}

	@Override
	public void reset() {
		super.reset();
		xVel = 0;
		yVel = 0;
	}
}

File: core/src/main/kotlin/net/brenig/pixelescape/game/entity/EntityMoving.kt
package net.brenig.pixelescape.game.entity

import net.brenig.pixelescape.game.data.GameMode

abstract class EntityMoving : Entity() {

    protected open var xVel: Float = 0F
    protected open var yVel: Float = 0F

    protected fun checkMaxVelocity(gameMode: GameMode) {
        xVel = Math.min(xVel, gameMode.maxEntitySpeed)
        yVel = Math.min(yVel, gameMode.maxEntitySpeed)
    }

    fun setVelocity(xVel: Float, yVel: Float) {
        this.xVel = xVel
        this.yVel = yVel
    }

    /**
     * moves the entity
     *
     *
     * note: does note check velocity limits
     *
     * @param delta time passed
     */
    protected fun move(delta: Float) {
        xPos += xVel * delta
        yPos += yVel * delta
    }

    override fun reset() {
        super.reset()
        xVel = 0f
        yVel = 0f
    }
}
